---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown



```{r }
library(httpuv)
library(devtools)
install_github("genomicsclass/GSE5859Subset", force = TRUE)
```

Here we have gene expression dataset
```{r }
library(GSE5859Subset)
data(GSE5859Subset)
dim(geneExpression)
```

We have RNA expression measurements for 8793 genes from blood taken from 24 individuals. We need information about the individuals to make statistical test. we will create subset from the data
```{r }
dim(sampleInfo)
head(sampleInfo)
sampleInfo$group
```

One of the columns, filenames, permits us to connect the rows of this table to the columns of the measurement table.
```{r }
match(sampleInfo$filename, colnames(geneExpression))
```

Finally, we have a table describing the features 
```{r }
dim(geneAnnotation)
head(geneAnnotation)
```

```{r }
head(match(geneAnnotation$PROBEID,rownames(geneExpression)))
```

How many samples were processed on 2005-06-27?
```{r }
library(dplyr)
sample <- filter(sampleInfo, date == "2005-06-27")
nrow(sample)
# or sum(sampleInfo$date=="2005-06-27")
```

How many of the genes represented in this particular technology are on chromosome Y?
```{r }
sum(geneAnnotation$CHR=="chrY", na.rm =TRUE)
```
na.rm =TRUE because some features are controls and have NA in the CHR column.

What is the log expression value of the for gene ARPC1A on the one subject that was measured on 2005-06-10?
```{r }
gene <- which(geneAnnotation$SYMBOL=="ARPC1A")
data <- which(sampleInfo$date=="2005-06-10")
geneExpression[gene, date]
```

Use the function apply() to find the median value of each column. What is the median value of these values?
```{r }
medians <- apply(geneExpression, 2, median)
median(medians)
```

Write a function that takes a vector of values e and a binary vector group coding two groups, and returns the p-value from a t-test: t.test( e[group==1], e[group==0])$p.value.

Now define g to code cases (1) and controls (0) like this g <- factor(sampleInfo$group)

Next use the function apply() to run a t-test for each row of geneExpression and obtain the p-value.

What is smallest p-value among all these t-tests?
```{r }
function_p <- function(e,group){
  x <- e[group==1]
  y <- e[group==0]
  return(t.test(x,y)$p.value)
}
g <- factor(sampleInfo$group)
pvalue <- apply(geneExpression,1,function_p, group=g)
min(pvalue)
```

We want to compare genes that are expressed different between cases and control by comparing the averages.
g: cases (1) and controls (0)
```{r }
library(GSE5859Subset)
data(GSE5859Subset)
g <- sampleInfo$group
```

we will take one feature out of 8000. here we will take feature 25 and we will make qq plot 
```{r }
e <- geneExpression[25,]
library(rafalib)
mypar(1,2)
qqnorm(e[g==1])
qqline(e[g==1])
qqnorm(e[g==0])
qqline(e[g==0])
```
The plots look like normal distribution and noe we will make t-tests
```{r }
t.test(e[g==1], e[g==0])
```
The p-value is not less than 0.05, so they are not significantly different 

Now we will make function to do the test for all features 
```{r }
myttest <- function(x) t.test(x[g==1],x[g==0],var.equal=TRUE)$p.value
myttest(geneExpression[25]) # for feature 25
pvals <- apply(geneExpression,1,mytest)
length(pvals)
```
```{r }
# Custom t-test function with check for the number of observations
t_test_function <- function(e, group) {
  group_1 <- e[group == 1]
  group_0 <- e[group == 0]

  # Check if there are enough observations in both groups
  if (length(group_1) < 2 || length(group_0) < 2) {
    # Return NA if there are not enough observations
    return(NA)
  }

  # Perform the t-test
  p_value <- t.test(group_1, group_0, var.equal = TRUE)$p.value
  return(p_value)
}
pvals <- apply(geneExpression,1,mytest)
length(pvals)
```
lets check genes that have p-value less than 0.05
```{r }
sum(pvals <= 0.05)
```
We don't think this is a good statistical method to say a conclusion about the genes

Lets create matrix using Monte Carlo simulation and the null is true for all data, no relationship
```{r }
m <- nrow(geneExpression)
n <- ncol(geneExpression)
randomData <-matrix(rnorm(n*m), m,n)
nullpvals <- apply(randomData, 1,myttest)
sum(nullpvals<=0.05)
```
We say that 420 are significantly different when it should be zero.
So p-values are random variables, we can't just report genes with p-value less that 0.05
Under the null hypothesis p-values follow a uniform distribution.

## Inference in Practice Exercises #1


```{r }
set.seed(1)
library(downloader)
url = "https://raw.githubusercontent.com/genomicsclass/dagdata/master/inst/extdata/femaleControlsPopulation.csv"
filename = "femaleControlsPopulation.csv"
if (!file.exists(filename)) download(url,destfile=filename)
population = read.csv(filename)
pvals <- replicate(1000,{
  control = sample(population[,1],12)
  treatment = sample(population[,1],12)
  t.test(treatment,control)$p.val
})
head(pvals)
hist(pvals)
```
What proportion of the p-values is below 0.05?
```{r }
mean(pvals < 0.05)
```

## Inference in Practice Exercises #2

What proportion of the p-values are below 0.01?
```{r }
mean(pvals < 0.01)
```

## Inference in Practice Exercises #3
Assume you are testing the effectiveness of 20 diets on mice weight. For each of the 20 diets, you run an experiment with 10 control mice and 10 treated mice. Assume the null hypothesis that the diet has no effect is true for all 20 diets and that mice weights follow a normal distribution with mean 30 grams and a standard deviation of 2 grams. Run a Monte Carlo simulation for one of these studies:
```{r }
cases = rnorm(10,30,2)
controls = rnorm(10,30,2)
t.test(cases,controls)$p.value
```

Now run a Monte Carlo simulation imitating the results for the experiment for all 20 diets. If you set the seed at 100, set.seed(100), and use the same code as above inside a call to replicate(), how many of the p-values (number not proportion) are below 0.05?
```{r }
set.seed(100)
pvals <- replicate(20,{
  cases = rnorm(10,30,2)
  controls = rnorm(10,30,2)
  t.test(cases,controls)$p.value
})
sum(pvals<=0.05)
```

## Inference in Practice Exercises #4

Set the seed at 100 again, set.seed(100), run the code from the previous question 1,000 times, and save the number of times the p-value is less than or equal to 0.05 for each of the 1,000 instances.

What is the average of these 1,000 numbers? Note that this is the expected number of tests (out of the 20 we run) that we will reject when the null is true.
```{r }
set.seed(100)
B =1000
pvalless = replicate(B,{
  pvals <- replicate(20,{
  cases = rnorm(10,30,2)
  controls = rnorm(10,30,2)
  t.test(cases,controls)$p.value
  })
  sum(pvals<=0.05)
})
mean(pvalless)
```

## Inference in Practice Exercises #5

Note that what the answer to question #4 says is that on average, we expect some p-value to be 0.05 even when the null is true for all diets.

Using the same simulation data from the question above, for what proportion of the 1,000 replicates do we reject the null hypothesis at least once (more than 0 false positives)?
```{r }
mean(pvalless>0)
```

## Quiz

```{r }
RNGkind("Mersenne-Twister", "Inversion", "Rejection")
```

Suppose you plan to run an experiment screening a panel of 30,000 small molecules to determine which ones increase expression of a fluorescent reporter gene. In untreated cells, the reporter gene expression follows a normal distribution with a mean of 8 units and a standard deviation of 2 units. There will be 100 untreated control wells, and each of the 30,000 molecules will be tested in 10 technical replicates. You want to simulate the experiment to figure out how many hits would come out of your screen if the null hypothesis is true for all 30,000 cases.
```{r }
set.seed(3)
ctrl = rnorm(100, 8, 2)
```

This example code simulates 10 technical replicates for one compound for which the null hypothesis is true:
```{r }
# use this code block inside your replicate() call
expt = rnorm(10, 8, 2)
t.test(ctrl, expt)$p.value
```

Now set the seed to 4 and use replicate() to simulate 30,000 tests for which the null hypothesis is true. The example code for one compound should go inside your replicate() call. Note that each test will compare the same ctrl vector to a new simulated experimental vector.
```{r }
set.seed(4)
B= 30000
func <- replicate(B,{
  expt = rnorm(10, 8, 2)
  t.test(ctrl, expt)$p.value
})
```

## Question 1
Make a histogram for p-values 
```{r }
hist(p_values, main = "Histogram of P-values", xlab = "P-value", col = "skyblue", border = "black")

```
## Question 2
What proportion of tests have a p-value below 0.05?

```{r }
mean(func <0.05)
```

## Question 3
Since this simulation assumes that the null distribution is true for all compounds, any results that have a p-value below a given cutoff will be false positives.

How many compounds have a p-value below 0.05?
```{r }
sum(func <0.05)
```

## Question 4
If the p-value cutoff is lowered to 0.001, how many false positive results are there?
```{r }
sum(func <0.001)
```

## Question 5-9 setup
Assume you are testing the effectiveness of 30 drugs on the white blood cell count of mice. For each of the 30 drugs you run an experiment with 5 control mice and 5 treated mice. Assume the null hypothesis that the drug has no effect is true for all 30 drugs and that white blood cell counts follow a normal distribution with mean 7.5 units and a standard deviation of 2.5 units.

We will analyze the number of significant p-values expected by chance under the null distribution.

## Question 5
Set the seed to 28, then run a Monte Carlo simulation for one of these studies by randomly generating white blood cell counts for the cases and controls. Use a t-test to compute the p-value for this simulated study.
What is the p-value for the one simulated study?
```{r }
set.seed(28)
cases = rnorm(5,30,2)
controls = rnorm(5,30,2)
t.test(cases,controls)$p.value
```

## Question 6
Now run a Monte Carlo simulation imitating the results for the experiment for all 30 drugs. Set the seed to 51, set.seed(51), then use your code from the previous question inside of replicate().

How many of the 30 simulated p-values are below 0.05?
```{r }
set.seed(51)
rep = replicate(30,{
  cases = rnorm(5,30,2)
  controls = rnorm(5,30,2)
  t.test(cases,controls)$p.value
})
sum(rep < 0.05)
```

## Question 7
Set the seed to 100, then repeat the simulated experiment 1000 times by using your code from the previous question inside a second replicate() loop. For each experiment, save the number of simulated p-values below 0.05.

What is the average of the counts of p-values below 0.05?

```{r }
set.seed(100)
B =1000
pvalless = replicate(B,{
  pvals <- replicate(30,{
  cases = rnorm(5,30,2)
  controls = rnorm(5,30,2)
  t.test(cases,controls)$p.value
  })
  sum(pvals<=0.05)
})
mean(pvalless)
```

## Question 8 

```{r }
hist(pvalless)
```