---
title: "EDA for Highthroughput"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Volcano plots

```{r }
library(genefilter)
library(GSE5859Subset)
data(GSE5859Subset)
g <- factor(sampleInfo$group)
results <- rowttests(geneExpression,g)
pvals <- results$pvalue
```

And we also generate p-values from a dataset for which we know the null is true:
```{r }
m <- nrow(geneExpression)
n <- ncol(geneExpression)
randomData <- matrix(rnorm(n*m),m,n)
nullpvals <- rowttests(randomData,g)$p.value
```

```{r }
plot(results$dm,-log10(results$p.value),
     xlab="Effect size",ylab="- log (base 10) p-values")
```

## p-value Histograms

```{r }
library(rafalib)
mypar(1,2)
hist(nullpvals,ylim=c(0,1400))
hist(pvals,ylim=c(0,1400))
```
## Data boxplots and histograms

```{r }
#devtools::install_github("genomicsclass/GSE5859")
library(Biobase)
library(GSE5859) 
data(GSE5859)
ge <- exprs(e) #ge for gene expression
ge[,49] <- ge[,49]/log2(exp(1)) #imitate error
```

```{r }
library(rafalib)
mypar(1,1)
boxplot(ge,range=0,names=1:ncol(e),col=ifelse(1:ncol(ge)==49,1,2))
```

```{r }
qs <- t(apply(ge,2,quantile,prob=c(0.05,0.25,0.5,0.75,0.95)))
matplot(qs,type="l",lty=1)
```

```{r }
mypar(1,1)
shist(ge,unit=0.5)
```

## MA plot
```{r }
x <- ge[,1]
y <- ge[,2]
mypar(1,2)
plot(x,y)
plot((x+y)/2,x-y)
```

```{r }
sd(y-x)
```

## Plots Exercises #1

```{r }
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("SpikeInSubset")
```

```{r }
library(SpikeInSubset)
data(mas133)
```

```{r }
e <- exprs(mas133)
plot(e[,1],e[,2],main=paste0("corr=",signif(cor(e[,1],e[,2]),3)),cex=0.5)
k <- 3000
b <- 1000 #a buffer
polygon(c(-b,k,k,-b),c(-b,-b,k,k),col="red",density=0,border="red")
```
What proportion of the points are inside the box?
```{r }
#We simply want to know which genes are below k for both samples:
mean(e[,1]<k & e[,2]<k)
```

## Plots Exercises #2

```{r }
plot(log2(e[,1]),log2(e[,2]),main=paste0("corr=",signif(cor(log2(e[,1]),log2(e[,2])),2)),cex=0.5)
k <- log2(3000)
b <- log2(0.5)
polygon(c(b,k,k,b),c(b,b,k,k),col="red",density=0,border="red")
```

## Plots Exercises #3

Make an MA-plot:
```{r }
e <- log2(exprs(mas133))
plot((e[,1]+e[,2])/2,e[,2]-e[,1],cex=0.5)
```

The two samples we are plotting are replicates (they random samples from the same batch of RNA). The correlation of the data was 0.997 in the original scale, 0.96 in the log-scale. High correlations are sometimes confused for evidence of replication. But replication implies we get very small difference between the observations, which is better measured with distance or differences.

What is the standard deviation of the log ratios for this comparison?
```{r }
#Here we simply compute the standard deviaton of the difference
sd(e[,2]-e[,1])
```

## Plots Exercises #4

How many fold changes above 2 do we see? Note that these measures of log (base 2) of expression so a fold change of 2 translates into a difference, in absolute value, of 1.
```{r }
#This are log2 measurements so a fold-change of 2 relates to differences of 1 (in absolute value). We then simply count the occurrences:

sum(abs(e[,2]-e[,1])>1)
```
